projects/driver_controls/src/ads1015.c

```diff
@@ -0,0 +1,169 @@
+      break;
+    case ADS1015_CHANNEL_2:
+      reg.data[0] = ADS1015_CONFIG_MUX_AIN2(reg.data[0]);
+      break;
+    case ADS1015_CHANNEL_3:
+      reg.data[0] = ADS1015_CONFIG_MUX_AIN3(reg.data[0]);
+      break;
+  }
+
+  status_ok_or_return(prv_write(s_i2c_port, ADS1015_CONFIG_REGISTER, reg.data));
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode ads1015_init(I2CPort i2c_port, GPIOAddress address) {
+  s_i2c_port = i2c_port;
+
+  // Configure the given GPIO address as a conversion ready pin
+  GPIOSettings gpio_settings = { .direction = GPIO_DIR_IN, .alt_function = GPIO_ALTFN_NONE };
+  InterruptSettings it_settings = { INTERRUPT_TYPE_INTERRUPT, INTERRUPT_PRIORITY_NORMAL };
+
+  gpio_init_pin(&address, &gpio_settings);
+  gpio_it_register_interrupt(&address, &it_settings, INTERRUPT_EDGE_FALLING,
+                              prv_interrupt_handler, NULL);
+
+  // Reset the internal registers to their default values
+  uint8_t reset = ADS1015_RESET_BYTE;
+  status_ok_or_return(i2c_write(s_i2c_port, 0, &reset, 1));
+
+  ADS1015Register reg;
+
+  // Read config register and write the proper configuration settings
+  status_ok_or_return(prv_read(s_i2c_port, ADS1015_CONFIG_REGISTER, reg.data));
+
+  reg.data[0] = ADS1015_CONFIG_MODE_CONT(reg.data[0]);
+  reg.data[1] = ADS1015_CONFIG_COMP_QUE_FOUR(reg.data[1]);
+  reg.data[1] = ADS1015_CONFIG_DR_128_SPS(reg.data[1]);
+
+  status_ok_or_return(prv_write(s_i2c_port, ADS1015_CONFIG_REGISTER, reg.data));
+
+  // Conversion ready must be initialized by setting the MSB of the HI_THRESH and LO_THRESH to 1
+  // and 0 respectively
+  reg.data[0] = ADS1015_HI_THRESH_RDY;
+  status_ok_or_return(prv_write(s_i2c_port, ADS1015_HI_THRESH_REGISTER, reg.data));
+  reg.data[0] = ADS1015_LO_THRESH_RDY;
+  status_ok_or_return(prv_write(s_i2c_port, ADS1015_LO_THRESH_REGISTER, reg.data));
+
+  // Initialize interrupt callbacks
+  for (uint8_t i = 0; i < NUM_ADS1015_CHANNELS; i++) {
+    s_interrupts[i].callback = NULL;
+    s_interrupts[i].context = NULL;
+    s_interrupts[i].reading = 0;
+  }
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode ads1015_register_callback(ADS1015Channel channel,
+                                      ADS1015Callback callback, void *context) {
+  status_ok_or_return(prv_channel_valid(channel));
+
+  s_interrupts[channel].callback = callback;
+  s_interrupts[channel].context = context;
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode ads1015_read_raw(ADS1015Channel channel, int16_t *reading) {
+  status_ok_or_return(prv_channel_valid(channel));
+
+  if (reading == NULL) {
+    return status_code(STATUS_CODE_EMPTY);
+  }
+
+  *reading = s_interrupts[channel].reading;
+  return STATUS_CODE_OK;
+}
+
+StatusCode ads1015_read_converted(ADS1015Channel channel, int16_t *reading) {
+  int16_t raw_reading;
+
+  StatusCode ret = ads1015_read_raw(channel, &raw_reading);
+
+  if (ret != STATUS_CODE_OK) {
+    return ret;
+  }
+
+  *reading = (raw_reading * ADS1015_REFERENCE_VOLTAGE)/2047;
```

Comment:
From what I remember, this family of chips has a PGA that sets the full scale voltage. If you're doing conversions, it should be against that.